cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,578 @@
+
+### Why a CRD in autoscaling.x-k8s.io and not a core k8s api?
+
+Buffers are not a required concept in each of the clusters and what they offer
+will depend on the autoscaling capabilities of the cluster. Adding them as a CRD
+in autoscaling space will make it clear and will allow for releasing it
+independently from the core k8s.
+
+## Features to be added in next iterations
+
+The initial iteration will cover basic use cases and will allow to simplify
+balloon pod/deployment management and allow for reduction of scheduler related
+overhead of preemption. But once we launch the initial integration there are
+potential next steps:
+
+*   Buffer dedicated for a workload - make it possible to create a buffer that 
+    would be dedicated for a workload and that can be "filled in" by it to limit
+    over provisioning if the user knows the max capacity needed by their
+    workload.
+*   To keep up with kubernetes changes we will add DRA support by adding a list
+    of resource claim templates that should be added to each chunk of the
+    buffer. This way it will be possible to define not only memory and cpu
+    capacity, but also any capacity that is possible to express with
+    ResourceClaim. This is not a part of the initial scope, because it would
+    require more implementation work within the cluster autoscaler in order to
+    support virtual pods with DRA resources.
+*   More features that further help users to manage the buffers:
+    *   Buffer schedule - option to define a buffer that is active only during
+        parts of the day (outside of these hours the buffer is ignored as if it
+        did not exist). Example usecase: I set up a buffer for the test jobs to
+        schedule quickly on the cluster. Tests are mostly run during business
+        hours of my organization and I do not want to pay for the extra capacity
+        at night or over the weekend.
+*   Buffer dedicated for workload: mark the buffer as dedicated for pods defined
+    by PodSelector. It would count the pods matching the selector as filling in
+    the buffer space. Example usecase: I want to keep buffer for scaling, but I
+    know the usual maximum needed for my workload (beyond it normal scale up
+    would apply)
+
+## User Stories (Optional)
+
+### Story 1
+
+The user has a deployment of pods that are serving user traffic. In order to
+better contain the traffic spikes I want to keep a buffer of a size of 10% of
+the deployment size so that when the HPA scales my deployment the new pods are
+started faster. Additionally this speeds up the feedback loop of the metrics
+allowing the HPA also to faster provide next scaling decisions.
+
+Example buffer:
+
+```
+apiversion: autoscaling.x-k8s.io/v1alpha1
+kind: NodeAutoscalingBuffer
+metadata:
+  name: my-deployment-buffer
+  namespace: my-namespace
+spec:
+  replicaBasedBufferSpec:
+    targetRef:
+      apiVersion: apps/v1
+      kind: Deployment
```

Comment:
I wonder -- does it make sense to support daemonsets here, too? There are cases where daemonsets can't safely upgrade when there is no room on the node, so this could force some space on each node.